### PR TITLE
Async tasks no delete

### PR DESCRIPTION
**Pull request by sblack-usu**
### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|
| NEW BUGS | 4 |
| NEW VULNERABILITIES | 0 |
| NEW SECURITY HOTSPOTS | 0 |
| NEW DUPLICATED LINES DENSITY | 0.375234521575985 |
| NEW CODE SMELLS | 6 |

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3971](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-3971)

File: django_irods/views.py
Comment (hyi) [inline django_irods/views.py:193]: I am thinking in the case of anonymous user, i.e., when a user does not login and try to download a public resource, is ```request.user.username``` an empty string? Would that cause potential issues for us? What do you think? @sblack-usu

Comment (sblack-usu) [reply django_irods/views.py:193]: Preparing downloads for anonymous users has not been addressed yet.  We'll need that implementation before this can be merged.

File: hs_core/migrations/0053_auto_20200826_1629.py
Comment (hyi) [inline hs_core/migrations/0053_auto_20200826_1629.py:19]: default cannot be 'pending' any more now that you removed 'pending' from the choices list.

File: hs_core/models.py
Comment (hyi) [inline hs_core/models.py:4408]: Have to change default to 'progress' now that pending status is removed.

File: hs_core/migrations/0053_auto_20200826_1629.py
Comment (sblack-usu) [reply hs_core/migrations/0053_auto_20200826_1629.py:19]: nice catch

File: hs_core/models.py
Comment (sblack-usu) [reply hs_core/models.py]: updated

**Comment by hyi** [comment]
@sblack-usu ```@task_postrun.connect``` does not appear to work in my local development environment, so task status is never updated but stay in progress status forever even though the task has completed successfully. Does this branch work in your local development environment?

File: hs_core/views/__init__.py
Comment (sblack-usu) [inline hs_core/views/__init__.py]: why is this handled differently than an ajax request?

Comment (hyi) [reply hs_core/views/__init__.py]: @sblack-usu This is for backward compatibility with delete resource REST API call but now that I think about it, REST API delete resource must have its own view function, i.e., not go through this view function. If that is indeed the case, I can remove this conditional check and handling. Let me know what you think.

File: hs_core/task_utils.py
Comment (sblack-usu) [inline hs_core/task_utils.py:69]: @hyi - I'm not following why we're not creating a task notification for anonymous users.  I don't see why we couldn't just follow the same workflow for anonymous users and just clear anonymous user TaskNotification entries after a day or two.  To take that further, I don't see why we couldn't clear TaskNotifications for all users after a day or two.  

Treating anonymous users differently as authenticated users within the workflows complicates the logic quite a bit and if we can treat all users as the same, we should.

File: hs_core/views/__init__.py
Comment (sblack-usu) [inline hs_core/views/__init__.py:106]: I see this if authenticated block several times.  It should be put in a method and called as needed.

File: hs_core/task_utils.py
Comment (hyi) [reply hs_core/task_utils.py:69]: @sblack-usu I did it this way after careful thought. To support anonymous users, the endpoint such as get_task() cannot be protected with login required meaning bots/hackers could hit that endpoint repeatedly creating a large number of junk entries in our task notification model if we just use get_or_create to create task entries without differentiating between anonymous users and authenticated users. The bots/hackers could also "legitimately" download public resources repeatedly which inflates a large number of junk entries into our model as well. In addition, creating a clean up celery task to clean up entries would require timestamp comparison (hence requiring adding a new timestamp field) and set up an threshold to determine when a task entry should be cleared up, which is also more code and logic. Also I think the logic I added for handling this is mostly encapsulated in the get_or_create_task_notification function, so I think it should be OK.

File: hs_core/views/__init__.py
Comment (hyi) [reply hs_core/views/__init__.py:106]: Good catch and will do.

File: hs_core/task_utils.py
Comment (sblack-usu) [reply hs_core/task_utils.py:77]: I would think extra TaskNotification entries are the least of our problems if bots/spam are repeatedly hitting the download endpoints.  Serving the downloads are much more likely to hurt performance.  A simple TaskNotification cleanup cron job probably should be implemented for authenticated users as well.

Comment (hyi) [reply hs_core/task_utils.py:77]: Agreed serving the downloads are much more likely to hurt performance than the extra entries in the model, but it is still a problem which I think is better to avoid. I initially thought a cleanup cron job probably should be implemented for authenticated users as well, but I changed my mind after some more thought. Each authenticated user has a clean all icon to click to clear all notification entries associated with the user, so proactively clearing up notification entries on behalf of users does not seem like a good idea to me.

**Comment by sblack-usu** [comment]
@hyi - the changes for anonymous download works, thanks for getting that sorted out.  I noticed an issue with zipped downloads (not bags).  Getting a 500 when trying to download a file/folder zipped.  Taking a look now.

**Comment by sblack-usu** [comment]
I'm also seeing this in the logs when requesting a zipped download.
>postgis            | ERROR:  duplicate key value violates unique constraint "hs_core_tasknotification_task_id_key"
postgis            | DETAIL:  Key (task_id)=(0c83fe52-b5e8-4978-b9e4-95f1bb57526d) already exists.
postgis            | STATEMENT:  INSERT INTO "hs_core_tasknotification" ("username", "task_id", "name", "payload", "status") VALUES ('', '0c83fe52-b5e8-4978-b9e4-

**Comment by sblack-usu** [comment]
It appears to be an issue with capital letters in the name used for the zip name.  Just checked to see if this exists in production, and downloads worked fine in production.

**Review by sblack-usu — COMMENTED** [review]
@hyi - I found the issue for downloading zipped files with capital letters.  I approve the rest of the code, please take a look and approve the PR if it's satisfactory.  Looking forward to having this finished up.

**Review by hyi — APPROVED** [review]

**Comment by hyi** [comment]
@sblack-usu Just approved the PR. Glad you found and fixed the issue with zipped file download with capital letters. Looking forward to having this finished up also!

**Pull request merged**
